serial.ts

```diff
@@ -53,119 +81,291 @@ const kParityIndexMapping: ParityType[] =
+    this.transferInterface_ = findInterface(
+        this.device_,
+        this.polyfillOptions_.usbTransferInterfaceClass as number);
+    this.inEndpoint_ = findEndpoint(this.transferInterface_, 'in');
+    this.outEndpoint_ = findEndpoint(this.transferInterface_, 'out');
   }
 
   /**
    * a function that opens the device and claims all interfaces needed to
    * control and communicate to and from the serial device
-   * @return {Promise} A promise that will resolve when device is ready for
-   * communication
+   * @param {SerialOptions} options Optional object containing serial options
+   * @return {Promise<void>} A promise that will resolve when device is ready
+   * for communication
    */
-  public async open() {
+  public async open(options?: SerialOptions): Promise<void> {
+    this.serialOptions_ = {...kDefaultSerialOptions, ...options};
+    this.validateOptions();
+
     try {
       await this.device_.open();
       if (this.device_.configuration === null) {
         await this.device_.selectConfiguration(1);
       }
 
-      const transferInterface = this.getTransferInterface(this.device_);
-      if (!transferInterface) {
-        throw new Error('Unable to find data transfer interface.');
-      }
-      this.transferInterface_ = transferInterface;
-
-      const controlInterface = this.getControlInterface(this.device_);
-      if (!controlInterface) {
-        throw new Error('Unable to find control interface.');
+      await this.device_.claimInterface(this.controlInterface_.interfaceNumber);
+      if (this.controlInterface_ !== this.transferInterface_) {
+        await this.device_.claimInterface(
+            this.transferInterface_.interfaceNumber);
       }
 
-      await this.device_.claimInterface(controlInterface.interfaceNumber);
-      await this.device_.claimInterface(
-          this.transferInterface_.interfaceNumber);
-      await this.setOptions();
-      await this.device_.controlTransferOut({
-        'requestType': 'class',
-        'recipient': 'interface',
-        'request': kSetControlLineState,
-        'value': 0x01,
-        'index': controlInterface.interfaceNumber,
-      });
-      this.readable = new ReadableStream({start: this.readStart.bind(this)});
-      this.writable = new WritableStream({
-        write: this.write.bind(this),
-      });
+      await this.setLineCoding();
+      await this.setSignals({dtr: true});
+
+      this.readable = new ReadableStream(
+          new UsbEndpointUnderlyingSource(this.device_, this.inEndpoint_),
+          new ByteLengthQueuingStrategy({
+            highWaterMark: this.serialOptions_.buffersize,
+          }));
+      this.writable = new WritableStream(
+          new UsbEndpointUnderlyingSink(this.device_, this.outEndpoint_),
+          new ByteLengthQueuingStrategy({
+            highWaterMark: this.serialOptions_.buffersize,
+          }));
     } catch (error) {
+      if (this.device_.opened) {
+        await this.device_.close();
+      }
```

Comment:
Seems like we should set dtr back to false if we don't have the streams working.  Maybe just call close()?
Not sure how readable.cancel() this.writable.abort() calls would behave if you called close now though.

Reply:
Construction of the streams won't fail as currently implemented. I've updated the steps for `close()` however to consider the case where the streams have not yet been created.